src/types/complex-types/array.ts

```diff
@@ -350,6 +350,30 @@ export function array<IT extends IAnyType>(subtype: IT): IArrayType<IT> {
     return new ArrayType<IT>(`${subtype.name}[]`, subtype)
 }
 
+function buildObjectByIdMap(nodes: AnyNode[]): [Set<string>, Map<string, Array<AnyNode>> | null] {
+    // Creates a map of node by identifier value.
+    //  In theory, several nodes can have the same identifier, if the array contains different types, so every identifier is mapped to an array of nodes with the same values.
+    //  In practice this in probably a rare case, so we can live with the performance hit.
+    //
+    // If not all nodes have identifier, we can't use the map for lookups, so we return null.
```

Comment:
Love that you've thought through this!

I'd highly recommend capturing various edge cases in test (if we haven't already done so). Some cases that come to mind would be tests for arrays of:

1. all scalars,
2. `model` types without identifiers,
3. `model` types with identifiers,
4. union of two distinct `model` types with and without identifiers
5. union of scalar and `model` type (I believe MST supports this, but double check).